src/sql/workbench/api/node/mainThreadNotebookDocumentsAndEditors.ts

```diff
@@ -0,0 +1,367 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+'use strict';
+
+import * as sqlops from 'sqlops';
+import { extHostNamedCustomer } from 'vs/workbench/api/electron-browser/extHostCustomers';
+import { Disposable } from 'vs/base/common/lifecycle';
+
+import URI, { UriComponents } from 'vs/base/common/uri';
+import { IExtHostContext } from 'vs/workbench/api/node/extHost.protocol';
+import { IEditorService } from 'vs/workbench/services/editor/common/editorService';
+import { IEditorGroupsService } from 'vs/workbench/services/group/common/editorGroupsService';
+import { IInstantiationService } from 'vs/platform/instantiation/common/instantiation';
+import { ITextEditorOptions } from 'vs/platform/editor/common/editor';
+import { viewColumnToEditorGroup } from 'vs/workbench/api/shared/editor';
+
+import {
+	SqlMainContext, MainThreadNotebookDocumentsAndEditorsShape, SqlExtHostContext, ExtHostNotebookDocumentsAndEditorsShape,
+	INotebookDocumentsAndEditorsDelta, INotebookEditorAddData, INotebookShowOptions, INotebookModelAddedData
+} from 'sql/workbench/api/node/sqlExtHost.protocol';
+import { NotebookInputModel, NotebookInput } from 'sql/parts/notebook/notebookInput';
+import { INotebookService, INotebookEditor } from 'sql/services/notebook/notebookService';
+import { TPromise } from 'vs/base/common/winjs.base';
+
+class MainThreadNotebookEditor extends Disposable {
```

Comment:
This is based on text editor implementation but is new

Reply:
May I know, why this class is required as we already have NotebookEditor?

Reply:
Good question. This is matching the pattern of the MainThreadDocumentsAndEditors, where they have a separate wrapper to act as an API for the main thread. The reason is that we'll add some Main/Ext host specific methods such as the 'matches' method, but more importantly will add some handlers for extension host API calls in the future.